src/components/DeeplinkWrapper/index.website.js

```diff
@@ -1,45 +1,95 @@
+    // 2. There may be non-idempotent operations (e.g. create a new workspace), which obviously should not be executed again in the desktop app.
+    // So we need to wait until after sign-in and navigation are complete before starting the deeplink redirect.
+    if (Str.startsWith(window.location.pathname, Str.normalizeUrl(ROUTES.TRANSITION_BETWEEN_APPS))) {
+        App.beginDeepLinkRedirectAfterTransition();
+    } else {
+        App.beginDeepLinkRedirect();
+    }
+}
+function DeeplinkWrapper({children, isAuthenticated}) {
+    const [currentScreen, setCurrentScreen] = useState();
+    const [hasShownPrompt, setHasShownPrompt] = useState(false);
+    const removeListener = useRef();
+
+    useEffect(() => {
+        // If we've shown the prompt and still have a listener registered,
+        // remove the listener and reset its ref to undefined
+        if (hasShownPrompt && removeListener.current !== undefined) {
+            removeListener.current();
+            removeListener.current = undefined;
+        }
+
+        if (isAuthenticated === false) {
+            setHasShownPrompt(false);
+            Navigation.isNavigationReady().then(() => {
+                // Get initial route
+                const initialRoute = navigationRef.current.getCurrentRoute();
+                setCurrentScreen(initialRoute.name);
+
+                removeListener.current = navigationRef.current.addListener('state', (event) => {
+                    setCurrentScreen(Navigation.getRouteNameFromStateEvent(event));
+                });
+            });
+        }
+    }, [hasShownPrompt, isAuthenticated]);
     useEffect(() => {
         // According to the design, we don't support unlink in Desktop app https://github.com/Expensify/App/issues/19681#issuecomment-1610353099
         const isUnsupportedDeeplinkRoute = _.some([CONST.REGEX.ROUTES.UNLINK_LOGIN], (unsupportRouteRegex) => {
             const routeRegex = new RegExp(unsupportRouteRegex);
             return routeRegex.test(window.location.pathname);
         });
-
-        if (!isMacOSWeb() || isUnsupportedDeeplinkRoute || CONFIG.ENVIRONMENT === CONST.ENVIRONMENT.DEV) {
+        // Making a few checks to exit early before checking authentication status
+        if (!isMacOSWeb() || isUnsupportedDeeplinkRoute || CONFIG.ENVIRONMENT === CONST.ENVIRONMENT.DEV || hasShownPrompt) {
             return;
         }
+        // We want to show the prompt immediately if the user is already authenticated.
+        // Otherwise, we want to wait until the navigation state is set up
+        // and we know the user is on a screen that supports deeplinks.
+        if (isAuthenticated) {
+            promptToOpenInDesktopApp();
+            setHasShownPrompt(true);
+        } else {
+            // Navigation state is not set up yet, we're unsure if we should show the deep link prompt or not
+            if (currentScreen === undefined || isAuthenticated === false) {
```

Comment:
These conditions did not consider opening magic link on logged in user, and caused this issue https://github.com/Expensify/App/issues/32795